### PR TITLE
Copy logs in a task

### PR DESCRIPTION
**Pull request by VladimirSlavik**
Copy logs in a task. Split from #3547.

~~Still fighting with tests.~~

Depends on #3584

### CODE REVIEW

File: pyanaconda/modules/boss/installation.py [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+
+    def _relabel_files(self):
+        """Relabel the anaconda logs.
+
+        The files we've just copied could be incorrectly labeled, like hawkey.log:
+        https://bugzilla.redhat.com/show_bug.cgi?id=1885772.
+        """
+        util.execInSysroot("restorecon", ["-ir", ANACONDA_LOG_DIR])
+
+    def _copy_file(self, src, dest):
```
Comment (jkonecny12) [inline pyanaconda/modules/boss/installation.py:136]: I would rename this to `_copy_file_to_sysroot()` or something similar to be obvious that the paths pointed as dest argument are in the sysroot.

File: tests/unit_tests/pyanaconda_tests/modules/boss/test_copy_logs_task.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+    @patch("pyanaconda.modules.boss.installation.util.execInSysroot")
+    @patch("pyanaconda.modules.boss.installation.util.execWithRedirect")
+    @patch("pyanaconda.modules.boss.installation.flags")
+    def test_run_all(self, flags_mock, exec_wr_mock, exec_isr_mock, glob_mock):
+        """Test the log copying task."""
+        glob_mock.return_value = ["/tmp/ks-script-blabbityblah.log"]
+        flags_mock.nosave_logs = False
+        flags_mock.nosave_input_ks = False
+
+        with tempfile.TemporaryDirectory() as dest:
```
Comment (jkonecny12) [inline tests/unit_tests/pyanaconda_tests/modules/boss/test_copy_logs_task.py:36]: You can also use `tmp_path` fixture nicely here if you want:
https://docs.pytest.org/en/latest/reference/reference.html#std-fixture-tmp_path

File: tests/unit_tests/pyanaconda_tests/modules/boss/test_copy_logs_task.py [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+# Red Hat, Inc.
+#
+import os.path
+import tempfile
+import unittest
+from unittest.mock import call, patch
+from pyanaconda.modules.boss.installation import CopyLogsTask
+
+
+class CopyLogsTaskTest(unittest.TestCase):
```
Comment (jkonecny12) [inline tests/unit_tests/pyanaconda_tests/modules/boss/test_copy_logs_task.py:25]: I would also like to see `test_nosave_all` test here.

**Review by jkonecny12 — COMMENTED** [review]
Looks great to me otherwise. Great job :)

File: tests/unit_tests/pyanaconda_tests/modules/boss/test_copy_logs_task.py (hunk 2 above)
Comment (VladimirSlavik) [reply tests/unit_tests/pyanaconda_tests/modules/boss/test_copy_logs_task.py:36]: Thank you for the review! I'm still trying to make the tests work, so likely I'll change it more. I want to patch also the dir creation and do nothing to the filesystem.

File: tests/unit_tests/pyanaconda_tests/modules/boss/test_copy_logs_task.py (hunk 3 above)
Comment (VladimirSlavik) [reply tests/unit_tests/pyanaconda_tests/modules/boss/test_copy_logs_task.py:25]: Yes, but `nosave_all` disappears way before this code, so here it is just the two flags. One flag, one test, and that covers all code paths.

File: tests/unit_tests/pyanaconda_tests/modules/boss/test_copy_logs_task.py (hunk 2 above)
Comment (jkonecny12) [reply tests/unit_tests/pyanaconda_tests/modules/boss/test_copy_logs_task.py]: Maybe the best option would be to patch functions in the shutil and not your functions. You will test more code by that.

File: tests/unit_tests/pyanaconda_tests/modules/boss/test_copy_logs_task.py (hunk 3 above)
Comment (jkonecny12) [reply tests/unit_tests/pyanaconda_tests/modules/boss/test_copy_logs_task.py:23]: I know there are two flags and I was thinking about testing both set. The point is to not miss some conflict between the flags.

**Comment by VladimirSlavik** [comment]
I'm not really happy yet, the granularity of the tests does not match the task. But it works, so feel free to pick it apart.

File: pyanaconda/modules/boss/installation.py (hunk 1 above)
Comment (VladimirSlavik) [reply pyanaconda/modules/boss/installation.py:136]: Doing the same with the other function too.

File: pyanaconda/modules/boss/installation.py [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+
+    def run(self):
+        self._copy_logs()
+        self._copy_kickstart()
+        self._relabel_files()
+
+    def _copy_logs(self):
+        """Copy installation logs to the target system.
+
+        This code is based on the 99-copy-logs.ks post script which is replaced by this.
```
[the remark this answers is not in the document]
Comment (VladimirSlavik) [reply pyanaconda/modules/boss/installation.py:56]: Done.

File: pyanaconda/modules/boss/installation.py [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+
+    def _create_logs_directory(self):
+        """Create directory for Anaconda logs on the install target"""
+        util.mkdirChain(self._sysroot + ANACONDA_LOG_DIR)
+
+    def _copy_tmp_logs(self):
+        """Copy a number of log files from tmp"""
+        LOG_FILES_TO_COPY = [
+            "anaconda.log", "syslog", "X.log", "program.log", "packaging.log", "storage.log",
+            "ifcfg.log", "lvm.log", "dnf.librepo.log", "hawkey.log", "dbus.log"
```
Comment (rvykydal) [inline pyanaconda/modules/boss/installation.py:79]: This should be handled in a separate commit or PR, but we are not using ifcfg.log anymore. (Should have been removed from the post script in the log file removal https://github.com/rhinstaller/anaconda/commit/ba2ab17e7ff3e5db1030ed1a709056c2142329fc). Would require also removing ifcfg.log from related unit test.

File: pyanaconda/modules/boss/installation.py [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+        ]
+        for logfile in LOG_FILES_TO_COPY:
+            self._copy_file_to_sysroot("/tmp/" + logfile,
+                                       ANACONDA_LOG_DIR + logfile)
+        # copy also list of packages used for creating the installation media
```
Comment (rvykydal) [inline pyanaconda/modules/boss/installation.py:84]: To be pedantic this does not seem to belong to the function according to its name, but feel free to keep it here.

File: pyanaconda/modules/boss/installation.py [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+        with open(self._sysroot + ANACONDA_LOG_DIR + "journal.log", "w") as logfile:
+            util.execWithRedirect("journalctl", ["-b"], stdout=logfile)
+
+    def _chmod_logs(self):
+        """Set access bits"""
+        util.execInSysroot("bash", ["-c", "chmod 0600 " + ANACONDA_LOG_DIR + "*"])
+
+    def _copy_kickstart(self):
+        """Copy generated kickstart file"""
+        log.info("Copying generated kickstart file...")
```
Comment (rvykydal) [inline pyanaconda/modules/boss/installation.py:115]: I think we are actually copying original/input - not generated kickstart file here ?

File: tests/unit_tests/pyanaconda_tests/modules/boss/test_boss.py [hunk 8]
```diff
@@ -223,6 +224,17 @@ def test_collect_install_system_tasks(self, proxy_getter, handler_getter):
             ("B", "/task/4"),
         ]
 
+    @patch_dbus_publish_object
+    def test_install_with_tasks(self, publisher):
+        """Test install tasks - Subscription module in default state."""
```
Comment (rvykydal) [inline tests/unit_tests/pyanaconda_tests/modules/boss/test_boss.py:229]: "Subscription module..." does not seem to belong here.

**Review by rvykydal — COMMENTED** [review]
Other than the comments looks good to me, thank you!

File: pyanaconda/modules/boss/installation.py (hunk 7 above)
Comment (VladimirSlavik) [reply pyanaconda/modules/boss/installation.py:115]: Double oops. I just copied the thing that was in the post script, but that was wrong... added by me, earlier.

**Comment by VladimirSlavik** [comment]
Update: Fixed all simple and unambiguous review remarks.

File: tests/unit_tests/pyanaconda_tests/modules/boss/test_copy_logs_task.py (hunk 3 above)
Comment (VladimirSlavik) [reply tests/unit_tests/pyanaconda_tests/modules/boss/test_copy_logs_task.py:23]: Let's do it, I guess.

File: tests/unit_tests/pyanaconda_tests/modules/boss/test_copy_logs_task.py (hunk 2 above)
Comment (VladimirSlavik) [reply tests/unit_tests/pyanaconda_tests/modules/boss/test_copy_logs_task.py]: Well, yes and no. The way tests are written now, filesystem is not touched. If I patched directly `copyfile` and `copytree`, I'd need also `os.path.exists`, and then we're at 8 patches per test. That sounds wrong.

**Comment by VladimirSlavik** [comment]
Update: Added the test for both nosaves "activated".

File: pyanaconda/modules/boss/boss.py [hunk 9]
```diff
@@ -141,3 +143,11 @@ def set_locale(self, locale):
         log.info("Setting locale of all modules to %s.", locale)
         super().set_locale(locale)
         self._module_manager.set_modules_locale(locale)
+
+    def copy_logs_with_task(self):
+        """Copy logs with an installation task.
+
+        :return: a DBus path of an installation task
```
Comment (poncovka) [inline pyanaconda/modules/boss/boss.py:150]: `an installation task`

File: pyanaconda/modules/boss/installation.py [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+    def run(self):
+        self._copy_logs()
+        self._copy_kickstart()
+        self._relabel_files()
+
+    def _copy_logs(self):
+        """Copy installation logs to the target system"""
+        log.info("Copying logs from the installation environment...")
+
+        if flags.nosave_logs:
```
Comment (poncovka) [inline pyanaconda/modules/boss/installation.py:57]: You should use the `conf` object. I guess you wanted to start with a task and implement the configuration options later. That's not possible. Flags are not synchronized or shared between modules, so they are valid only in the UI process. That's why we have created the configuration files. This flag will be always set to False in a DBus module. I think you can create the configuration options in another pull request and then rebase this one.

File: pyanaconda/modules/boss/installation.py [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+            log.warning("Installation logs will not be saved to the installed system due to the "
+                        "nosave option.")
+        else:
```
Comment (poncovka) [inline pyanaconda/modules/boss/installation.py:60]: Please, call `return` in the `if` branch above and remove `else`.

File: pyanaconda/modules/boss/installation.py [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+            self._copy_dnf_debugdata()
+            self._copy_post_script_logs()
+            self._dump_journal()
+            self._chmod_logs()
+
+            log.info("Done.")
+
+    def _create_logs_directory(self):
+        """Create directory for Anaconda logs on the install target"""
+        util.mkdirChain(self._sysroot + ANACONDA_LOG_DIR)
```
Comment (poncovka) [inline pyanaconda/modules/boss/installation.py:73]: Please, call `join_paths` to join the paths.

File: pyanaconda/modules/boss/installation.py [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+
+    def _copy_tmp_logs(self):
+        """Copy a number of log files from tmp"""
+        LOG_FILES_TO_COPY = [
+            "anaconda.log", "syslog", "X.log", "program.log", "packaging.log", "storage.log",
```
Comment (poncovka) [inline pyanaconda/modules/boss/installation.py:78]: It might be more readable to have one item per line, but it's up to you.

File: pyanaconda/modules/boss/installation.py [hunk 14]
```diff
@@ -0,0 +1,151 @@
...
+                                       ANACONDA_LOG_DIR + os.path.basename(logfile))
+
+    def _dump_journal(self):
+        """Dump journal from the installation environment"""
+        with open(self._sysroot + ANACONDA_LOG_DIR + "journal.log", "w") as logfile:
+            util.execWithRedirect("journalctl", ["-b"], stdout=logfile)
+
+    def _chmod_logs(self):
+        """Set access bits"""
+        util.execInSysroot("bash", ["-c", "chmod 0600 " + ANACONDA_LOG_DIR + "*"])
```
Comment (poncovka) [inline pyanaconda/modules/boss/installation.py:111]: I think you can call `os.chmod` or, at least, call `chmod` directly without `bash`.

File: pyanaconda/modules/boss/installation.py [hunk 15]
```diff
@@ -0,0 +1,151 @@
...
+        """Copy input kickstart file"""
+        log.info("Copying input kickstart file...")
+
+        if flags.nosave_input_ks:
+            log.warning("Input kickstart will not be saved to the installed system due to the "
+                        "nosave option.")
+        else:
+            self._copy_file_to_sysroot("/run/install/ks.cfg",
+                                       "/root/original-ks.cfg")
+            log.info("Done.")
```
Comment (poncovka) [inline pyanaconda/modules/boss/installation.py:123]: I am not sure if the `Done.` messages are necessary. How the logs look like with them?

File: pyanaconda/modules/boss/installation.py [hunk 16]
```diff
@@ -0,0 +1,151 @@
...
+        :type sysroot: str
+        """
+        super().__init__()
+        self._sysroot = sysroot
+
+    @property
+    def name(self):
+        return "Copy installation logs"
+
+    def run(self):
```
Comment (poncovka) [inline pyanaconda/modules/boss/installation.py:48]: Please, add a docstring with a detailed description of the steps.

File: tests/unit_tests/pyanaconda_tests/modules/boss/test_copy_logs_task.py [hunk 17]
```diff
@@ -0,0 +1,191 @@
...
+# Red Hat, Inc.
+#
+import unittest
+from unittest.mock import call, patch
+from pyanaconda.modules.boss.installation import CopyLogsTask
+
+
+class CopyLogsTaskTest(unittest.TestCase):
+    @patch("pyanaconda.modules.boss.installation.glob.glob")
+    @patch("pyanaconda.modules.boss.installation.util.execInSysroot")
```
Comment (poncovka) [inline tests/unit_tests/pyanaconda_tests/modules/boss/test_copy_logs_task.py:25]: You shouldn't use `execInSysroot`in the task. It means that the task doesn't always use the given `sysroot` argument. Please, use an alternative.

**Review by poncovka — CHANGES_REQUESTED** [review]
The flags won't work as expected.

File: tests/unit_tests/pyanaconda_tests/modules/boss/test_copy_logs_task.py [hunk 18]
```diff
@@ -0,0 +1,191 @@
...
+    @patch("pyanaconda.modules.boss.installation.os.path.exists")
+    def test_copy_tree_to_sysroot(self, exists_mock, copytree_mock):
+        """Test _copy_tree_to_sysroot"""
+        task = CopyLogsTask("/somewhere")
+
+        exists_mock.return_value = True
+        task._copy_tree_to_sysroot("/some/source", "/another/destination/")
+        exists_mock.assert_called_with("/some/source")
+        copytree_mock.assert_called_with("/some/source", "/somewhere/another/destination/",
+                                         dirs_exist_ok=True)
```
Comment (jkonecny12) [inline tests/unit_tests/pyanaconda_tests/modules/boss/test_copy_logs_task.py:183]: I would like to separate each parameters here to a new line.

**Review by jkonecny12 — COMMENTED** [review]
I agree on what @poncovka and @rvykydal wrote here. Aside that it looks good to me.

File: tests/unit_tests/pyanaconda_tests/modules/boss/test_boss.py [hunk 19]
```diff
@@ -223,6 +224,17 @@ def test_collect_install_system_tasks(self, proxy_getter, handler_getter):
...
+    @patch_dbus_publish_object
+    def test_install_with_tasks(self, publisher):
+        """Test CopyLogsWithTask."""
+        task_path = self.interface.CopyLogsWithTask()
+        task_proxy = check_task_creation(task_path, publisher, CopyLogsTask)
+        task = task_proxy.implementation
+        assert task.name == "Copy installation logs"
+
+        task = self.module.copy_logs_with_task()
+        assert task.name == "Copy installation logs"
```
[the remark this answers is not in the document]
Comment (VladimirSlavik) [reply tests/unit_tests/pyanaconda_tests/modules/boss/test_boss.py:236]: I am actually not sure what the second test would be. Maybe it's redundant?

[the remark this answers is not in the document]
Comment (poncovka) [reply tests/unit_tests/pyanaconda_tests/modules/boss/test_boss.py:236]: I should have been more specific. You don't have to test the `CopyLogsWithTask` DBus method and the `copy_logs_with_task` Python method. The test of the Python method is redundant.

**Review by poncovka — APPROVED** [review]
It looks great. Thanks!

**Review by jkonecny12 — APPROVED** [review]
Looks great to me. Thanks!

**Review by rvykydal — APPROVED** [review]
Looks good to me.

**Comment by AdamWill** [comment]
This looks to have broken live image composes in Rawhide:
https://pagure.io/releng/failed-composes/issue/2752
https://koji.fedoraproject.org/koji/taskinfo?taskID=75833053

**Comment by VladimirSlavik** [comment]
Thanks for the news. Looks like the error happens in `_copy_kickstart`, and likely is caused by missing `/root`. It's a pity there is no syslog or journal from the build, the original exception is lost.

**Comment by AdamWill** [comment]
the passed-through anaconda.log does have this:

    08:31:53,879 INF threading: Thread Failed: AnaInstallThread (139789092066880)
    08:31:53,880 DBG exception: running handleException
    08:31:53,883 CRT exception: Traceback (most recent call last):

      File "/usr/lib64/python3.10/site-packages/pyanaconda/installation_tasks.py", line 527, in run_task
        sync_run_task(self._task_proxy)

      File "/usr/lib64/python3.10/site-packages/pyanaconda/modules/common/task/__init__.py", line 46, in sync_run_task
        task_proxy.Finish()

      File "/usr/lib/python3.10/site-packages/dasbus/client/handler.py", line 444, in _call_method
        return self._get_method_reply(

      File "/usr/lib/python3.10/site-packages/dasbus/client/handler.py", line 477, in _get_method_reply
        return self._handle_method_error(error)

      File "/usr/lib/python3.10/site-packages/dasbus/client/handler.py", line 497, in _handle_method_error
        raise exception from None

    dasbus.error.DBusError: [Errno 2] No such file or directory: '/mnt/sysroot/root/original-ks.cfg'


    During handling of the above exception, another exception occurred:


    Traceback (most recent call last):

      File "/usr/lib64/python3.10/site-packages/pyanaconda/threading.py", line 275, in run
        threading.Thread.run(self)

      File "/usr/lib64/python3.10/threading.py", line 946, in run
        self._target(*self._args, **self._kwargs)

      File "/usr/lib64/python3.10/site-packages/pyanaconda/installation.py", line 424, in run_installation
        queue.start()

      File "/usr/lib64/python3.10/site-packages/pyanaconda/installation_tasks.py", line 311, in start
        item.start()

      File "/usr/lib64/python3.10/site-packages/pyanaconda/installation_tasks.py", line 311, in start
        item.start()

      File "/usr/lib64/python3.10/site-packages/pyanaconda/installation_tasks.py", line 311, in start
        item.start()

      File "/usr/lib64/python3.10/site-packages/pyanaconda/installation_tasks.py", line 496, in start
        self.run_task()

      File "/usr/lib64/python3.10/site-packages/pyanaconda/installation_tasks.py", line 530, in run_task
        if errorHandler.cb(e) == ERROR_RAISE:

      File "/usr/lib64/python3.10/site-packages/pyanaconda/errors.py", line 219, in cb
        raise NonInteractiveError("Non interactive installation failed: %s" % exn)

    pyanaconda.errors.NonInteractiveError: Non interactive installation failed: [Errno 2] No such file or directory: '/mnt/sysroot/root/original-ks.cfg'

**Comment by VladimirSlavik** [comment]
Yes, I saw that, that is the useless part. It's the top process getting a dbus message that one of the children tracebacked. The real exception is in journal and syslog. The only useful bit is the exception message, which is one if the ambiguous ones anyway.

The juicy bits would be in the logs of the other process anyway... We modularized and suddenly no longer the same logs matter.
